digdag-cli/src/main/java/io/digdag/cli/SelfUpdate.java

```diff
@@ -100,28 +104,79 @@ private void selfUpdate(String version)
                         version, res.getStatus(), res.getStatusInfo(), res.readEntity(String.class)));
         }
 
+        boolean backgroundMove = false;
         Path path = Files.createTempFile("digdag-", ".bat");
-        try (InputStream in = res.readEntity(InputStream.class)) {
-            try (OutputStream out = Files.newOutputStream(path)) {
-                ByteStreams.copy(in, out);
+        try {
+            try (InputStream in = res.readEntity(InputStream.class)) {
+                try (OutputStream out = Files.newOutputStream(path)) {
+                    ByteStreams.copy(in, out);
+                }
+            }
+            path.toFile().setExecutable(true, false);
+            path.toFile().setReadable(true, false);
+
+            // try to move the file to make sure that the file is on the same file system
+            // with the destination so that enable move more likely works. This is also
+            // helpful if /tmp file system is mounted with noexec option on Linux.
+            Path near = dest.getParent().resolve(".digdag.selfupdate.bat");
+            try {
+                Files.move(path, near, REPLACE_EXISTING);
```

Comment:
In fact, this change is not for Windows. On UNIX, updating a file using atomic operation is necessary to not make a running java process crashed. Java process opens a jar file and keeps its file descriptor opened so that it reads class files lazily. Overriding the file durning running the process impacts this lazy loading.